linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala

```diff
@@ -327,125 +332,152 @@ final class Analyzer(config: CommonPhaseConfig, initial: Boolean,
 
   private def lookupClass(className: ClassName)(
       onSuccess: ClassInfo => Unit)(implicit from: From): Unit = {
-    lookupClassForLinking(className, Set.empty) {
+    workTracker.track(classLoader.lookupClass(className)) {
       case info: ClassInfo =>
         info.link()
         onSuccess(info)
 
-      case CycleInfo(cycle, _) =>
-        _errors += CycleInInheritanceChain(cycle, fromAnalyzer)
+      case CycleInfo(cycle, root) =>
+        assert(root == null, s"unresolved root: $root")
+        addError(CycleInInheritanceChain(cycle, fromAnalyzer))
     }
   }
 
-  private def lookupClassForLinking(className: ClassName,
-      knownDescendants: Set[LoadingClass] = Set.empty)(
-      onSuccess: LoadingResult => Unit): Unit = {
-
-    _classInfos.get(className) match {
-      case None =>
-        val loading = new LoadingClass(className)
+  private final class ClassLoader(implicit ec: ExecutionContext) {
```

Comment:
I originally isolated this into its own class to make `_classInfos` more isolated. Now I'm not so sure anymore it's really an improvement. WDYT?

Reply:
I'm not sure either. It doesn't seem substantially better or worse either way. What I do find a bit problematic is that this change combined in the same commit that makes it thread-safe makes things quite a bit blurry to me. It's unclear what has actually changed in terms of thread-safety. Is it just `_classInfos` becoming a `TrieMap`? If yes, can we put that in a separate commit?

Reply:
>  Is it just _classInfos becoming a TrieMap?

Ish, yes. But doing this also implies moving the actual loading code from `LodingClass` to its own method (`doLoad`) inside the `ClassLoader`. Essentially, we use the `_classInfos` as a synchronization point to chose the thread which is supposed to trigger class loading. Before this commit, that was part of the constructor of `LoadingClass`.

>  If yes, can we put that in a separate commit?

Not trivially, but I can attempt to split it into smaller refactoring steps if you think it would be helpful.

I have attempted to suggest an ordering in this comment, but it is non-trivial since with the thread-safety, we can also leverage futures more, so this makes other code paths (notably the new different code paths for lookupClassForLinking and lookupClass) manageable (whereas w/o futures, they'd need additional helpers).

Reply:
OK, I see. Let's keep it as is, then. It's not that hard to make sense of.